### ISSUE TITLE
Possible to explicitly specify output paths?

### ISSUE THREAD

**Issue by macandcheese** [middleman/middleman#2082]
### I'd like to be able to explicitly specify the output destination of directories during build

For example, I have a "/templates" folder in my source, that contains many sub directories, each of which contains template files. I want to change where the contents of that folder get built to.

For sanity, I like to keep this /templates folder in source directory, rather than moving all of its children "up a level" to my directory root. However, on build output, I would like the contents of "/templates" to be served to "/". I understand Middleman likes to map 1:1 but for larger projects it's unmanageable to keep all template files in a flat structure.

Is there a way to accomplish this? Thanks in advance.

**Comment by macandcheese** [comment]
So, using after_build, I'm able to manually move the contents. For example I want to move /es, /fr, etc. from /translations where they are built to the root directory, as well as move the contents of /templates up to the root, I do the following:

`after_build do |builder|
  # after build we move our translations folders up to the root of the project so we can access them as url roots
  # we use force incase a previous build's files remain, we'll overwrite with latest
  FileUtils.mv(Dir['build/translations/*'],'build/', {:force => true})
  # and we also move out template files up to the root so we can have cleaner urls
  FileUtils.mv(Dir['build/templates/*'],'build/', {:force => true})
  # then delete the folders in their original locations
  FileUtils.rm_rf('build/templates')
  FileUtils.rm_rf('build/translations')
end`

It would be nice to have this be a during-build option if possible to avoid extra steps during deploy!

**Comment by iwarner** [comment]
Hiya 

I do not understand why you want to move whats inside /templates though - they should be dynamic and utilised by proxy pages.

It seems to me you should create most of these as partials and then call them into say modules that already exist in the source. Check out my pattern here: https://github.com/iwarner/CodeBlender-Middleman/tree/master/source  and https://github.com/iwarner/CodeBlender-Middleman/tree/master/source/atom/badge I have loads of partials in the source folder that I simply call into the templates. Maybe we are doing something similar?

Check out : https://github.com/middleman/middleman/issues/697

Basically you can build an extension to utilise the internal thor actions and do what you need to do - either within Config or as a standalone lib.

also an example of a extension that uses this
https://github.com/bbtfr/middleman-bower/blob/master/lib/middleman-bower/commands.rb

Hope this helps a little at least

Ian

**Comment by macandcheese** [comment]
@iwarner Thanks for the response, maybe I didn't explain well. I think we do have a similar set up.

Here's my source
``
source
-partials
-layouts
-templates
--resources
---casestudy
----template.html.erb
----index.html.erb
---ebook
---etc
--blog
--etc
-javascripts
-images
-etc
``
So, I do build my dynamic pages via proxy pages helper - that all works fine. Where I run into trouble is getting the contents of /templates in source to output to /build ROOT so that my published paths don't contain /templates.

Further, I've found the need to roll my own translation system, and it works almost perfectly. Basically, on build, I'm manually creating looping through my partials, and layouts, and certain template files, copying those to new locations (basically /partials/translations, /translations/ etc.), with a folder per language, and then do a regex search and replace to replace instances of "en-US" with all languages I've defined. Again, this works fine.

However, I'm left with /translations/es, /translations/fr, /translations/etc in my source directory. Which is intended, as I don't want to litter my source directory with dozens of language folders.

After build, I use FileUtils to take the contents of /translations, and /templates that have been output into build directory, and move them all "up a level" to the root of the build directory, so the paths on deployed site for users are /fr vs /translations/fr, and /blog vs. /templates/blog

SO - now my build folder is totally "flat" - and my templates are all output as desired to root, however because I moved the files with FileUtils, MM's sitemap is not aware of the new structure, and as such middleman s3_sync can't upload this new structure. So, I'm manually calling

`import_path File.expand_path('build/*', app.root)`

which, upon looking at /__middleman, correctly maps my build (now, this means that my local sitemap has both versions of paths, which isn't ideal, but works)

BUT - the issue here is that things that are dependent on the MM sitemap during build process, like creating a sitemap with builder, aren't aware since the files are moved after build.

Looking through your code it seems we have a similar setup, it's just my desire to keep all templates within a folder in source to stay organized.
